### PR TITLE
Fix eviction logic in GetLruPages

### PR DESCRIPTION
**Pull request by JuanCarlosGI**
Change GetLruPages logic to have two queues, filled in pages; one with the oldest content, and one starting in the middle.

### CODE REVIEW

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 1]
```diff
@@ -257,61 +257,149 @@ public Task<BoolResult> UnregisterAsync(Context context, IReadOnlyList<ContentHa
...
+        }
+
+        /// <summary>
+        /// Takes an enumerable and takes n=pageSize elements from it. Processess the elements with a query, and inserts
+        /// them into a priority queue. It will then continue yielding elements of the queue, until it has less than
+        /// n elements, at which point it will take another n elements from the original enumerable and repeat the
+        /// process, adding them to the same priority queue. This process repeats until the original enumerable has no
+        /// more elements.
+        /// </summary>
+        private static IEnumerable<T> QueryAndOrderInPages<T>(IEnumerable<T> original, int pageSize, Comparer<T> comparer, Func<List<T>, IEnumerable<T>> query)
```
Comment (ljcollins25) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: >QueryAndOrderInPages [](start = 38, length = 20)

Maybe move to NuCacheCollectionUtilities

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 2]
```diff
@@ -257,61 +257,149 @@ public Task<BoolResult> UnregisterAsync(Context context, IReadOnlyList<ContentHa
...
-            while (priorityQueue.Count != 0)
-            {
-                yield return GetPriorityPage(priorityQueue, subPageSize);
+                yield return queue.Top;
+                queue.Pop();
             }
         }
 
-        private IReadOnlyList<TItem> GetPriorityPage<TItem>(PriorityQueue<TItem> queue, int pageSize)
+        private static IEnumerable<T> MergeOrdered<T>(IEnumerable<T> items1, IEnumerable<T> items2, IComparer<T> comparer)
```
Comment (ljcollins25) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: Maybe move to NuCacheCollectionUtilities

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 3]
```diff
@@ -257,61 +257,149 @@ public Task<BoolResult> UnregisterAsync(Context context, IReadOnlyList<ContentHa
...
 
-            var operationContext = new OperationContext(context);
+            Func<List<ContentHashWithLastAccessTimeAndReplicaCount>, IEnumerable<ContentHashWithLastAccessTimeAndReplicaCount>> query =
+                page => _localLocationStore.GetEffectiveLastAccessTimes(operationContext, page.SelectList(v => new ContentHashWithLastAccessTime(v.ContentHash, v.LastAccessTime))).ThrowIfFailure();
 
-            var subPageSize = Math.Max(1, pageSize / 10);
+            // We make sure that we we select a set of the newer content, to ensure that we at least look at newer content to see if it should be
+            // evicted first due to having a high number of replicas. We do this by looking at the start as well as at middle of the list.
+            var localOldest = QueryAndOrderInPages(contentHashesWithInfo.Take(contentHashesWithInfo.Count / 2), pageSize, comparer, query);
+            var localMid = QueryAndOrderInPages(contentHashesWithInfo.Skip(contentHashesWithInfo.Count / 2), pageSize, comparer, query);
```
Comment (ljcollins25) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: >Skip [](start = 70, length = 4)

Skip isn't optimized for lists. Maybe just have QueryAndOrderInPages take a list with start and end offset.

File: Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs [hunk 4]
```diff
@@ -257,61 +257,149 @@ public Task<BoolResult> UnregisterAsync(Context context, IReadOnlyList<ContentHa
...
 
-            // Content is retrieved in batches of size, PageSize, and then returned in pages of subPageSize=PageSize/10 as long as the queue has >= PageSize elements.
-            // This ensures that we have PageSize amount of lookahead
-            // NOTE: We use RandomSplitInterleave to ensure we select a random set of the newer content. This ensure we at least look at newer content to see if it should be
-            // evicted first due to having a high number of replicas
-            foreach (var page in contentHashesWithInfo.RandomSplitInterleave().GetPages(pageSize))
+            var mergedEnumerables = MergeOrdered(localOldest, localMid, comparer);
+            return GetPages(mergedEnumerables, subPageSize);
```
Comment (ljcollins25) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: >GetPages [](start = 19, length = 8)

Just use the built-in GetPages method.

Comment (ljcollins25) [inline Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: >subPageSize [](start = 47, length = 11)

I don't think we need subpages anymore.

Comment (JuanCarlosGI) [reply Public/Src/Cache/ContentStore/Distributed/NuCache/TransitioningContentLocationStore.cs]: Is there one? I can't find it

**Pull request merged**